cpp/src/io/parquet/page_hdr.cu

```diff
@@ -540,24 +632,192 @@ void __launch_bounds__(decode_page_headers_block_size)
           }
           bs->page.kernel_mask = kernel_mask_for_page(bs->page, bs->ck);
         } else {
+          error[warp_id] |=
+            static_cast<kernel_error::value_type>(decode_error::INVALID_PAGE_HEADER);
           bs->cur = bs->end;
         }
+        if (index_out >= 0 and index_out < max_num_pages) { page_info[index_out] = bs->page; }
       }
-      index_out = shuffle(index_out);
-      if (index_out >= 0 && index_out < max_num_pages && lane_id == 0) {
-        page_info[index_out] = bs->page;
+      values_found = shuffle(values_found);
+      warp.sync();
+    }
+    if (lane_id == 0 and error[warp_id] != 0) { set_error(error[warp_id], error_code); }
+  }
+}
+
+/**
+ * @brief Kernel for counting the number of page headers from the specified column chunks
+ *
+ * @param[in] chunks Device span of column chunks
+ * @param[out] error_code Pointer to the error code for kernel failures
+ */
+CUDF_KERNEL void __launch_bounds__(count_page_headers_block_size)
+  count_page_headers_kernel(cudf::device_span<ColumnChunkDesc> chunks,
+                            kernel_error::pointer error_code)
+{
+  auto constexpr num_warps_per_block = decode_page_headers_block_size / cudf::detail::warp_size;
```

Comment:
Same contents as `decode_page_headers_kernel` except only page counting related code is left and the rest is removed